### PR TITLE
feat: Add 3ID Lifecycle page

### PR DESCRIPTION
**Pull request by oed**
Describes the lifecycle of 3IDs.

### CODE REVIEW

File: docs/authentication/3id-did/lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+## 3ID DID provider lifecycle
+
+![asdf](../../images/3id-lifecycle.png)
+
+### Creating a 3ID
+To create a 3ID the *3id-did-provider* accepts an `authSecret` and an `authId`. It roughly follows the following algorithm.
+
+1. We have an `authSecret A` that is used to create `did:key:A` using the [`key-did-provider-ed25519`](https://github.com/ceramicnetwork/key-did-provider-ed25519) provider
+1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is a new `authSerect` the loaded *AuthLink* will be empty
```
Comment (ukstv) [inline docs/authentication/3id-did/lifecycle.md:15]:
```suggestion
1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is a new `authSecret` the loaded *AuthLink* will be empty
```

File: docs/authentication/3id-did/lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+### Revoking an `authSecret`
+Assume that we have `authSecret A` and `authSecret B` associated to our 3ID, and we want to revoke `authSecret A`. Lets refer to the current seed as `seed 1`.
+
+1. Randomly generate a new seed, `seed 2`
+1. Three keys (`controller`, `signing`, and `encryption`) are derived from `seed 2`, and the 3ID stream for the 3ID DID document is updated
+1. The `seed 1` is encrypted to `seed 2` and stored in the 3ID Keychain, so that we can use it to decrypt any message encrypted to the previous version of the 3ID document
+1. `seed 2` is encrypted to `authSecret B` but **not** to `authSecret A`
+1. The encrypted `seed 2` is stored in the 3ID keychain
+
+After these steps are taken, `authSecret A` can no longer be used to authenticate to the 3ID.
```
Comment (ukstv) [inline docs/authentication/3id-did/lifecycle.md:45]: Note here: revocation can only happen if at least one authSecret is available. Before revoking an old authSecret, one should add a new one.

That was a source of confusion few times.

**Review by ukstv — APPROVED** [review]

File: docs/authentication/3id-did/lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# 3ID Lifecycle
+
+This page describes the complete lifecycle of a new 3ID, how the DID document is created, how keys are managed, and how 3ID connect uses these building blocks. Most of the logic described here is implemented in the [3id-did-provider](https://github.com/ceramicstudio/js-3id-did-provider) package.
```
Comment (michaelsena) [inline docs/authentication/3id-did/lifecycle.md:3]:
```suggestion
This page describes the complete lifecycle of a [3ID DID](./method.md), including how the DID document is created, how keys are managed, and how [3ID Connect](./3id-connect.md) uses these building blocks. Most of the logic described here is implemented in the [3id-did-provider](./provider.md) package.
```

File: docs/authentication/3id-did/lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+The two core components to the 3ID lifecycle is the [3ID DID method](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-79/CIP-79.md) and the [3ID Keychain](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-20/CIP-20.md) which is an IDX definition for storing encrypted key material.
```
Comment (michaelsena) [inline docs/authentication/3id-did/lifecycle.md:5]:
```suggestion
The two core components to the 3ID lifecycle are the [3ID DID method](./method.md) and the [3ID Keychain](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-20/CIP-20.md) which is a standard for storing encrypted key material in a DID's [IDX](../../tools/identity/idx.md).
```

File: docs/authentication/3id-did/lifecycle.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+## 3ID DID provider lifecycle
+
+![asdf](../../images/3id-lifecycle.png)
```
Comment (michaelsena) [inline docs/authentication/3id-did/lifecycle.md:9]:
```suggestion
![3ID DID Provider Lifecycle Diagram](../../images/3id-lifecycle.png)
```

File: docs/authentication/3id-did/lifecycle.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+### Creating a 3ID
+To create a 3ID the *3id-did-provider* accepts an `authSecret` and an `authId`. It roughly follows the following algorithm.
+
+1. We have an `authSecret A` that is used to create `did:key:A` using the [`key-did-provider-ed25519`](https://github.com/ceramicnetwork/key-did-provider-ed25519) provider
```
Comment (michaelsena) [inline docs/authentication/3id-did/lifecycle.md:14]:
```suggestion
1. We have an `authSecret A` that is used to create `did:key:A` using the [`key-did-provider-ed25519`](../key-did/provider.md) provider
```

File: docs/authentication/3id-did/lifecycle.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is a new `authSerect` the loaded *AuthLink* will be empty
+1. A seed for the 3ID is randomly generated, three keys (`controller`, `signing`, and `encryption`) are derived from the seed, and the stream for the 3ID DID document is created
+1. The seed is encrypted to `did:key:A`
+1. The encrypted seed is put into the 3ID Keychain, which is an IDX record associated with the 3ID from step 3
+1. The *AuthLink* document is updated to contain the 3ID from step 2.
+
+### Loading an existing 3ID
+If an `authSecret` has already been associated with a 3ID then this 3ID will be loaded instead of a new one created.
+
+1. `authSecret A` is used to create `did:key:A` using the [`key-did-provider-ed25519`](https://github.com/ceramicnetwork/key-did-provider-ed25519) provider.
```
Comment (michaelsena) [inline docs/authentication/3id-did/lifecycle.md:24]:
```suggestion
1. `authSecret A` is used to create `did:key:A` using the [`key-did-provider-ed25519`](../key-did/provider.md) provider.
```

File: docs/authentication/3id-did/lifecycle.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+### Creating a 3ID
+To create a 3ID the *3id-did-provider* accepts an `authSecret` and an `authId`. It roughly follows the following algorithm.
+
+1. We have an `authSecret A` that is used to create `did:key:A` using the [`key-did-provider-ed25519`](https://github.com/ceramicnetwork/key-did-provider-ed25519) provider
+1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is a new `authSerect` the loaded *AuthLink* will be empty
+1. A seed for the 3ID is randomly generated, three keys (`controller`, `signing`, and `encryption`) are derived from the seed, and the stream for the 3ID DID document is created
+1. The seed is encrypted to `did:key:A`
+1. The encrypted seed is put into the 3ID Keychain, which is an IDX record associated with the 3ID from step 3
+1. The *AuthLink* document is updated to contain the 3ID from step 2.
```
Comment (stbrody) [inline docs/authentication/3id-did/lifecycle.md:19]:
```suggestion
1. The *AuthLink* document is updated to contain the 3ID from step 3.
```

Comment (stbrody) [reply docs/authentication/3id-did/lifecycle.md:19]: step 2 is for making the authlink that is being updated, the 3id is created in step 3, right?

File: docs/authentication/3id-did/lifecycle.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+### Loading an existing 3ID
+If an `authSecret` has already been associated with a 3ID then this 3ID will be loaded instead of a new one created.
+
+1. `authSecret A` is used to create `did:key:A` using the [`key-did-provider-ed25519`](https://github.com/ceramicnetwork/key-did-provider-ed25519) provider.
+1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is an existing `authSerect` the loaded *AuthLink* contain a 3ID
+1. Using the 3ID from the previous step we can deterministically load the 3ID keychain
+1. Decrypt the seed in the 3ID Keychain that was encrypted to `did:key:A`
```
Comment (stbrody) [inline docs/authentication/3id-did/lifecycle.md:27]: why is the seed needed if you've already loaded the 3id stream?

Comment (stbrody) [reply docs/authentication/3id-did/lifecycle.md:27]: I think this is only needed in order to do updates to the 3id document, i.e. to add or revoke authSecrets - is that right?  I guess the seed is used to generate the did:key that is the `controller` of the 3id stream?  So I guess you need to authenticate as that did:key before being able to make updates the 3id stream?
Assuming I got that all right I don't think it's very clear here in the text and could use clarifying.

Comment (stbrody) [reply docs/authentication/3id-did/lifecycle.md:27]: Probably could add some steps to both `Adding a new authSecret` and `Revoking an authSecret` related to authenticating as the did:key that is the 3id stream controller, by using the seed.

File: docs/authentication/3id-did/lifecycle.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+1. The encrypted `seed 2` is stored in the 3ID keychain
+
+After these steps are taken, `authSecret A` can no longer be used to authenticate to the 3ID.
+
+## Usage in 3ID Connect
+A natural question at this point is: *Where does these authSecrets come from?* In 3ID Connect normal blockchain wallets are used to create these secrets in the following way.
+
+1. A signature of some static message is requested from the users wallet, e.g. MetaMask
+1. If the user accepts a signature is returned
+1. The entropy of the signature is now used as an `authSecret` (the hash of the signature), in addition 3ID Connect will use the wallet AccountID as the `authId`
```
Comment (stbrody) [inline docs/authentication/3id-did/lifecycle.md:52]: It's never been clear to me what the word "entropy" means in this context.  Maybe that's a standard term in the cryptography world?  Would be nice to be more clear here about what we mean by that for people who may not be familiar with the terminology

**Review by stbrody — CHANGES_REQUESTED** [review]

File: docs/authentication/3id-did/lifecycle.md (hunk 5 above)
Comment (oed) [reply docs/authentication/3id-did/lifecycle.md:9]: lol, ty

File: docs/authentication/3id-did/lifecycle.md (hunk 10 above)
Comment (oed) [reply docs/authentication/3id-did/lifecycle.md:52]: > Maybe that's a standard term in the cryptography world?

It's very standard yes. I'll see what I can do.

File: docs/authentication/3id-did/lifecycle.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
+# 3ID Lifecycle
+
+This page describes the complete lifecycle of a [3ID DID](./method.md), including how the DID document is created, how keys are managed, and how [3ID Connect](./3id-connect.md) uses these building blocks. Most of the logic described here is implemented in the [3id-did-provider](./provider.md) package.
+
+The two core components to the 3ID lifecycle are the [3ID DID method](./method.md) and the [3ID Keychain](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-20/CIP-20.md) which is a standard for storing encrypted key material in a DID's [IDX](../../tools/identity/idx.md).
+
+## 3ID DID provider lifecycle
+Every 3ID has a *seed* which is the secret used to update the 3ID, sign messages, and decrypt messages as the 3ID. The 3ID DID provider manages this *seed* using *authSecrets*. Each `authSecret` can be used to authenticate to the 3ID and thus get access to the *seed*.
```
Comment (oed) [inline docs/authentication/3id-did/lifecycle.md:8]: @stbrody added this paragraph. Hopefully it makes things more clear.

File: docs/authentication/3id-did/lifecycle.md [hunk 12]
```diff
@@ -1,27 +1,28 @@
...
 To create a 3ID the *3id-did-provider* accepts an `authSecret` and an `authId`. It roughly follows the following algorithm.
 
-1. We have an `authSecret A` that is used to create `did:key:A` using the [`key-did-provider-ed25519`](https://github.com/ceramicnetwork/key-did-provider-ed25519) provider
-1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is a new `authSerect` the loaded *AuthLink* will be empty
-1. A seed for the 3ID is randomly generated, three keys (`controller`, `signing`, and `encryption`) are derived from the seed, and the stream for the 3ID DID document is created
+1. We have an `authSecret A` that is used to create `did:key:A` using the [`key-did-provider-ed25519`](../key-did/provider.md) provider
+1. A deterministic TileDocument, referred to as *AuthLink*, is created. If this is a new `authSecret` the loaded *AuthLink* will be empty
+1. A seed for the 3ID is randomly generated; two keys (`signing`, and `encryption`), and a `did:key:management-1` are derived from the seed. A stream for the 3ID DID document is created with `did:key:management-1` as the *controller*, the `signing` and `encryption` keys are stored in the streams content
```
Comment (stbrody) [inline docs/authentication/3id-did/lifecycle.md:17]: The 'signing' and 'encryption' keys that are put in the stream contents are the *public* keys for signing and encryption of an asymmetric key pair, right?  They can't be symmetric keys or you wouldn't want to put them in the stream contents which is public.  Would be nice to clarify that

**Review by stbrody — COMMENTED** [review]
one question about the signing and encryption keys stored in the stream contents. Otherwise looks good!

File: docs/authentication/3id-did/lifecycle.md (hunk 12 above)
Comment (oed) [reply docs/authentication/3id-did/lifecycle.md:17]: good point!

**Review by stbrody — APPROVED** [review]
LGTM!

**Pull request merged**